File: sender/sender.go
// Package sender POSTs the collected DiscoveredAssets to the CryptoSight
// probe ingest endpoint with Bearer token authentication.
//
// Retry policy: up to 3 attempts with exponential back-off (2 s, 4 s).
// Any non-2xx final response is treated as a hard error so the caller can
// log it and decide whether to retry on the next cron tick.
package sender

import (
        "bytes"
        "encoding/json"
        "fmt"
        "io"
        "log"
        "net/http"
        "strings"
        "time"

        "github.com/cryptosight/probe/types"
)

const (
        maxAttempts    = 3
        initialBackoff = 2 * time.Second
)

// Send batches assets into a single POST to <endpoint>/probes/ingest.
// Pass snifferStats as non-nil only from the passive sniffer flush path;
// active-scan callers should pass nil.
// It returns the server's accepted/rejected counts on success.
func Send(endpoint, apiKey, probeVersion, hostname string, assets []types.DiscoveredAsset, snifferStats *types.SnifferStats) (*types.IngestResponse, error) {
        url := strings.TrimRight(endpoint, "/") + "/probes/ingest"

        body := types.IngestRequest{
                Assets:       assets,
                ProbeVersion: probeVersion,
                Hostname:     hostname,
                SnifferStats: snifferStats,
        }

        payload, err := json.Marshal(body)
        if err != nil {
                return nil, fmt.Errorf("marshalling ingest payload: %w", err)
        }

        var lastErr error
        backoff := initialBackoff

        for attempt := 1; attempt <= maxAttempts; attempt++ {
                resp, err := post(url, apiKey, payload)
                if err == nil {
                        return resp, nil
                }
                lastErr = err
                if attempt < maxAttempts {
                        log.Printf("WARN: ingest attempt %d/%d failed: %v — retrying in %s", attempt, maxAttempts, err, backoff)
                        time.Sleep(backoff)
                        backoff *= 2
                }
        }

        return nil, fmt.Errorf("ingest failed after %d attempts: %w", maxAttempts, lastErr)
}

func post(url, apiKey string, payload []byte) (*types.IngestResponse, error) {
        req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(payload))
        if err != nil {
                return nil, fmt.Errorf("building request: %w", err)
        }
        req.Header.Set("Content-Type", "application/json")
        req.Header.Set("Authorization", "Bearer "+apiKey)

        client := &http.Client{Timeout: 30 * time.Second}
        res, err := client.Do(req)
        if err != nil {
                return nil, fmt.Errorf("HTTP POST: %w", err)
        }
        defer res.Body.Close()

        respBody, _ := io.ReadAll(res.Body)

        if res.StatusCode < 200 || res.StatusCode >= 300 {
                return nil, fmt.Errorf("server returned %d: %s", res.StatusCode, strings.TrimSpace(string(respBody)))
        }

        var ingestResp types.IngestResponse
        if err := json.Unmarshal(respBody, &ingestResp); err != nil {
                // Non-JSON success body is tolerated.
                return &types.IngestResponse{}, nil
        }
        return &ingestResp, nil
}

File: sender.py
"""HTTP sender — POSTs discovered assets to the CryptoSight ingest endpoint.

Retry policy: up to 3 attempts with exponential back-off (2 s, 4 s).
Any non-2xx final response is raised as an exception.
"""

from __future__ import annotations

import logging
import time
from typing import Optional

import requests

from assets import DiscoveredAsset, SnifferStats

logger = logging.getLogger(__name__)

_MAX_ATTEMPTS = 3
_INITIAL_BACKOFF = 2.0


def send(
    endpoint: str,
    api_key: str,
    probe_version: str,
    hostname: str,
    assets: list[DiscoveredAsset],
    *,
    sniffer_stats: Optional[SnifferStats] = None,
    ssl_verify: bool = True,
) -> dict:
    """POST assets to <endpoint>/probes/ingest and return the parsed response."""
    url = endpoint.rstrip("/") + "/probes/ingest"

    payload: dict = {
        "assets": [a.to_dict() for a in assets],
        "probeVersion": probe_version,
        "hostname": hostname,
    }
    if sniffer_stats is not None:
        payload["snifferStats"] = sniffer_stats.to_dict()

    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {api_key}",
    }

    last_err: Optional[Exception] = None
    backoff = _INITIAL_BACKOFF

    for attempt in range(1, _MAX_ATTEMPTS + 1):
        try:
            resp = requests.post(
                url, json=payload, headers=headers, timeout=30, verify=ssl_verify
            )
            if resp.ok:
                try:
                    return resp.json()
                except Exception:
                    return {}
            last_err = RuntimeError(f"server returned {resp.status_code}: {resp.text[:200]}")
        except requests.RequestException as e:
            last_err = e

        if attempt < _MAX_ATTEMPTS:
            logger.warning(
                "WARN: ingest attempt %d/%d failed: %s — retrying in %.0fs",
                attempt, _MAX_ATTEMPTS, last_err, backoff,
            )
            time.sleep(backoff)
            backoff *= 2

    raise RuntimeError(f"ingest failed after {_MAX_ATTEMPTS} attempts: {last_err}")
